pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java

```diff
@@ -168,12 +168,13 @@ private ByteBuf getCompressedBatchMetadataAndPayload() {
         // Update the current max batch size using the uncompressed size, which is what we need in any case to
         // accumulate the batch content
         maxBatchSize = Math.max(maxBatchSize, uncompressedSize);
+        maxMessagesNum = Math.max(maxMessagesNum, numMessagesInBatch);
         return compressedPayload;
     }
 
     @Override
     public void clear() {
-        messages = new ArrayList<>();
+        messages = new ArrayList<>(maxMessagesNum);
```

Comment:
Can we use `message.clear()` instead of creating new ArrayList here?

Reply:
If using `message.clear()`, there will be some cpu overhead I think.  We must check if `messages.size == maxMessageNum`, and the clear() method will loop all the elements in the arrayList:
```
 public void clear() {
        modCount++;
        final Object[] es = elementData;
        for (int to = size, i = size = 0; i < to; i++)
            es[i] = null;
    }
```